### PR TITLE
Add: Some additional debugging information when chunk sizes are not as expected

### PR DESCRIPTION
**Pull request by LordAro**
## Motivation / Problem

Debugging SL errors is hard when they provide no information except "Invalid chunk size" with no context. Also when that message comes from one of 4 possible locations

## Description

Add some context

## Limitations

N/A

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/saveload/saveload.cpp [hunk 1]
```diff
@@ -694,6 +694,7 @@ int SlIterateArray()
 			return INT32_MAX;
 		}
 
+		int index;
```
Comment (nielsmh) [inline src/saveload/saveload.cpp:676]: Is this variable used?

**Review by nielsmh — COMMENTED** [review]
Is there any more information that is available that might be useful to include? Something like current chunk name perhaps?

File: src/saveload/saveload.cpp (hunk 1 above)
Comment (LordAro) [reply src/saveload/saveload.cpp:676]: Yes, see the following lines :p

**Comment by LordAro** [comment]
I've added what I can without altering function signatures and the like (only a couple have references to the chunk that's currently being loaded)

Have also converted usage to fmt instead of printf style. It's a bit ick due to the "extern" error in pool_func.hpp

File: src/core/pool_func.hpp [hunk 2]
```diff
@@ -14,7 +14,7 @@
 #include "mem_func.hpp"
 #include "pool_type.hpp"
 
-extern void NORETURN SlErrorCorruptFmt(const char *format, ...) WARN_FORMAT(1, 2);
+#include "../saveload/saveload.h" // SlErrorCorrupt
```
Comment (rubidium42) [inline src/core/pool_func.hpp]: Maybe create a new `saveload_error.h` with the `SlError` functions? Including that in here will at least not include all the file IO stuff.

File: src/saveload/saveload_error.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file saveload.h Functions/types related to errors from savegames. */
+
+#ifndef SAVELOAD_ERROR_HPP
+#define SAVELOAD_ERROR_HPP
+
+#include "../stdafx.h"
```
Comment (rubidium42) [inline src/saveload/saveload_error.hpp]: This include is not wanted in headers. Although I see a few headers that (erroneously I'd say) include it.

File: src/saveload/saveload_error.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * @tparam A The type of the first argument.
+ * @tparam Args The types of the other arguments.
+ * @note This function does never return as it throws an exception to
+ *       break out of all the saveload code.
+ */
+template <typename T, typename A, typename ... Args>
+static inline void NORETURN SlErrorCorruptFmt(const T &format, A first_arg, Args&&... other_args)
+{
+	/* The separate first_arg argument is added to aid overloading.
+	 * Otherwise the calls that do no need formatting will still use this function. */
```
Comment (rubidium42) [inline src/saveload/saveload_error.hpp]: There is no overloading of the function as far as I can see. Did you want to call this function `SlErrorCorrupt`? Or is it just sloppy-paste?

File: src/saveload/saveload.h [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #ifndef SAVELOAD_H
 #define SAVELOAD_H
 
+#include "saveload_error.hpp"
 #include "../fileio_type.h"
 #include "../fios.h"
 #include "../strings_type.h"
```
Comment (rubidium42) [inline src/saveload/saveload.h]: saveload_error.h includes strings_type, so it's not needed here anymore.

**Review by rubidium42 — COMMENTED** [review]
I like that you changed this over from printf to format!

File: src/saveload/saveload_error.hpp (hunk 4 above)
Comment (LordAro) [reply src/saveload/saveload_error.hpp]: Yup, straight copied from IConsolePrint :) I could rename this, or I could squash the args together...

File: src/saveload/saveload.h (hunk 5 above)
Comment (LordAro) [reply src/saveload/saveload.h]: Even though this header itself uses StringID?

Comment (rubidium42) [reply src/saveload/saveload.h]: Yes. It makes little sense to include something that has already been included. It only slows things up.
Maybe it's time for me to do the old... can I compile each header with only stdafx.h, and which includes can I remove from the headers spiel again. But that's a slow and tedious process.
Having said that, it's not a huge problem to leave it either. After all, with the header protection it should skip most of the file from being processed.

Comment (LordAro) [reply src/saveload/saveload.h]: https://github.com/include-what-you-use/include-what-you-use ?

Comment (rubidium42) [reply src/saveload/saveload.h]: The question is a bit... where to draw the line? Technically this file should then also include `bitmath_func.hpp`, `cassert`, 'memory' and `stdafx.h`.
All in all it doesn't really matter. It's just that less includes help a tiny bit in building performance. Either way is ultimately fine by me.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**